pycbc/io/hdf.py

```diff
@@ -402,47 +469,84 @@ class SingleDetTriggers(object):
+
+        premask : array of indices or boolean, optional
+           Array of used triggers
+
+        filter_rank : string, optional
+            The ranking, as defined by ranking.py to compare to
+            filter_threshold
+
+        filter_threshold: float, required if filter_rank is used
+            Threshold to filter the ranking values
+
+        chunksize : int , default 1e6
+            Size of chunks to read in for the filter_rank / threshold.
+        """
         logging.info('Loading triggers')
         self.trigs_f = HFile(trig_file, 'r')
         self.trigs = self.trigs_f[detector]
+        self.ntriggers = self.trigs['end_time'].size
         self.ifo = detector  # convenience attributes
         self.detector = detector
         if bank_file:
             logging.info('Loading bank')
             self.bank = HFile(bank_file, 'r')
         else:
-            logging.info('No bank file given to SingleDetTriggers')
             # empty dict in place of non-existent hdf file
             self.bank = {}
 
-        if premask is not None:
-            self.mask = premask
-        else:
-            self.mask = np.ones(len(self.trigs['end_time']), dtype=bool)
+        # Apply some masks to start off with - here we should try and apply
+        # them in the order which cuts most things earliest.
+        self.mask = premask
+
+        if filter_rank:
+            assert filter_threshold is not None
+            logging.info("Applying threshold of %.3f on %s",
+                         filter_threshold, filter_rank)
+            idx = self.trigs_f.select(
+                 lambda rank: rank > filter_threshold,
+                 derived={ranking.sngls_ranking_function_dict[filter_rank]:
+                          ranking.required_datasets[filter_rank]},
+                 indices_only=True,
+                 premask=self.mask,
+                 group=detector,
+                 chunksize=chunksize,
+            )
+            logging.info("%d triggers remain", idx.size)
+            # If self.mask already has values, need to take these into account:
+            self.and_masks(idx)
 
-        if veto_file:
-            logging.info('Applying veto segments')
-            # veto_mask is an array of indices into the trigger arrays
-            # giving the surviving triggers
-            logging.info('%i triggers before vetoes', self.mask.sum())
-            self.veto_mask, _ = events.veto.indices_outside_segments(
-                self.end_time, [veto_file],
-                ifo=detector, segment_name=segment_name)
-
-            idx = np.flatnonzero(self.mask)[self.veto_mask]
-            self.mask[:] = False
-            self.mask[idx] = True
-            logging.info('%i triggers remain after vetoes',
-                          len(self.veto_mask))
-
-        # FIXME this should use the hfile select interface to avoid
-        # memory and processing limitations.
         if filter_func:
```

Comment:
`pycbc_plot_singles_vs_params`, `pycbc_process_sngls` and `pycbc_fit_sngl_trigs` all use the `filter_func` option (from git grepping), I don't know whether all these are still needing to use it though

I think this could be sent to the select() function, but wanted to sort other parts first

Reply:
One complication here is where the filter_func uses derived datasets, or template bank parameters

Reply:
I'll look at this after this has been merged - there has already been quite a bit of scope creep of this PR